[Huang Hongyi] iP

Hongyi6328 opened the pull request:
# This is Hongyi's version of Duke

> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

Using this app, you can:

- Manage your time effectively
- Plan for your deadlines and schedules

Programming paradigms employed:
1. Object-oriented programming. For example, `TodoTask task = new TodoTask(taskTitle);`
2. Functional Programming. For example,
    ```Java
   private static final Function<String, Command> ADD_TODO_COMMAND_SUPPLIER = commandArgument -> {
        Command newCommand;
        try {
            String taskTitle = getTaskTitle(commandArgument);
            TodoTask task = new TodoTask(taskTitle);
            newCommand = new AddTodoCommand(task);
        } catch (DukeMissingTaskTitleException exception) {
            newCommand = new ErrorCommand(exception.getMessage());
        }
        return newCommand;
    };
    ```

**Features** include: :heart_eyes:
- [x] Add task
- [x] Delete
- [x] Find
- [ ] Mark

Commit:
Add Gradle support

Hongyi6328 committed:
Complete Level-1

Hongyi6328 committed:
Complete Level-2

Hongyi6328 committed:
Modify Duke.java

We try not to use that many classes by this time.

Hongyi6328 committed:
Complete Level-3

Hongyi6328 committed:
Add Level-4, error handling

Hongyi6328 committed:
Add command fetcher

Hongyi6328 committed:
Complete Level 5

Hongyi6328 committed:
Add the delete function

Hongyi6328 committed:
Add 3 subclasses: TodoTask, DeadlineTask, and EventTask

Hongyi6328 committed:
Complete A-TextUiTesting

Hongyi6328 committed:
Improve string representations of different tasks

Hongyi6328 committed:
Rectify the index system such that indices are more in common sense

Hongyi6328 committed:
Create packages

Hongyi6328 committed:
Create packages

Hongyi6328 committed:
Fix some minor bugs

Hongyi6328 committed:
Fix some bugs

Hongyi6328 committed:
Add more exception classes

Hongyi6328 committed:
Add a save function

Hongyi6328 committed:
Add a DateTimeFormatter

Hongyi6328 committed:
Merge pull request #1 from Hongyi6328/branch-Level-7

Add a save function

Hongyi6328 committed:
Resolve conflicts between master and branch-Level-8

Hongyi6328 committed:
Resolve conflicts between master and branch-Level-8

Hongyi6328 committed:
Resolve conflicts between master and branch-Level-8

Hongyi6328 committed:
Merge pull request #2 from Hongyi6328/branch-Level-8

Add a DateTimeFormatter

Hongyi6328 committed:
Add packages

Hongyi6328 committed:
Add More OOP

Hongyi6328 committed:
Merge pull request #3 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support

Hongyi6328 committed:
Fix some bugs

Hongyi6328 committed:
Add unit test cases

Hongyi6328 committed:
Release a jar file

Hongyi6328 committed:
Remove unused variables

Hongyi6328 committed:
Remove unused variables

Hongyi6328 committed:
Merge from Hongyi6328/branch-A-CodingStandard

Remove unused variables

Hongyi6328 committed:
Add Level 9

Hongyi6328 committed:
Merge pull request #5 from Hongyi6328/branch-Level-9

Add Level 9

Hongyi6328 committed:
Reconstruct the code

Hongyi6328 committed:
Reconstruct classes

Hongyi6328 committed:
Add JavaDoc

Hongyi6328 committed:
Merge from Hongyi6328/branch-A-JavaDoc

Add JavaDoc

Hongyi6328 committed:
Fix some bugs

Hongyi6328 committed:
Fix some bugs

Hongyi6328 committed:
Reorganise code and classes

Hongyi6328 committed:
Add checkstyle config

Hongyi6328 committed:
Check code style

Hongyi6328 committed:
Add more JavaDoc

Hongyi6328 committed:
Add more JavaDoc

Hongyi6328 committed:
Add more JavaDoc

Hongyi6328 committed:
Optmize imports and disable wildcard imports

Hongyi6328 committed:
Fix some bugs

Hongyi6328 committed:
Add more tests

t1mzzz commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Duke project template
+# duke.Duke project template
```
This might be a result of packaging duke. Should you change this to Duke project template?
```suggestion
# Duke project template
```

t1mzzz commented on README.md:
```diff
@@ -13,7 +13,7 @@ Prerequisites: JDK 11, update Intellij to the most recent version.
 1. Configure the project to use **JDK 11** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br>
    In the same dialog, set the **Project language level** field to the `SDK default` option.
-3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
+3. After that, locate the `src/main/java/duke.Duke.java` file, right-click it, and choose `Run duke.Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
```
Similar to the above, this might be a result of packaging duke. Should you change duke.Duke to Duke?
```suggestion
3. After that, locate the `src/main/java/duke.Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
```

t1mzzz commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,76 @@
+    private static final String FILE_WRITING_DELIMITER = "|";
+
+    protected String taskTitle;
+    protected boolean done;
```
Should this boolean variable sound more like a boolean?
```suggestion
    protected boolean isDone;
```

t1mzzz commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,273 @@
+        return newCommand;
+    };
+
+    private static final Function<String, Command> unknownCommandSupplier = commandArgument -> new UnknownCommand();
```
Perhaps this `static final` variable name can be changed to upper case, where each word is divided by "_"?

t1mzzz reviewed (commented):
Overall, the code is really structured and really well written. It is really impressive! There are some minor changes that need to be made, but mistakes were really hard to find throughout. One main issue I found was the very long class, namely Parser. Split some of these methods in the class and delete some redundant methods if possible. Finally, don't forget to add your PR description!

Hongyi6328 committed:
Refine code to follow conventions

Hongyi6328 committed:
Finish checkstyle configuration

Hongyi6328 committed:
Improve code quality

Hongyi6328 committed:
Add JavaFX

Hongyi6328 committed:
Build a better GUI

Hongyi6328 committed:
Add more powerful GUI features

Hongyi6328 committed:
Add varargs syntax sugar

Hongyi6328 committed:
Correct coding standard violations

Hongyi6328 committed:
Merge from branch-A-CheckStyle

Correct coding standard violations

Hongyi6328 committed:
Add MainWindow class. Separate UI and Logic

Hongyi6328 committed:
Modify DialogBox

Hongyi6328 committed:
Seperate CLI and GUI control classes

Hongyi6328 committed:
Change Ui class name to CliUi

Hongyi6328 committed:
Catch up commits from branch-Level-10

Jnwkm commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,38 @@
+ * An enum class that lists out all command types. This can be used to determine the runtime type of a command object.
+ */
+public enum CommandType {
+    UNKNOWN(""),
+    ERROR(null),
+    EXIT("bye"),
+    DISPLAY_LIST("list"),
+    MARK_DONE("mark"),
+    MARK_UNDONE("unmark"),
+    DELETE("delete"),
+    FIND("find"),
+    ADD_EVENT("event"),
+    ADD_TODO("todo"),
+    ADD_DEADLINE("deadline");
```
Like the uses of enum here!

Jnwkm commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+package duke.exception;
```
Like how you named each of your exceptions to make it clearer which exception is being returned!

Jnwkm reviewed (commented):
Looks good really follows the naming conventions well! Couldn't really find any problems here :)

Hongyi6328 committed:
Resolve conflicts with the master branch

Hongyi6328 committed:
Update CliUi.java

Hongyi6328 committed:
Merge from branch-Level-10

Branch level 10

Hongyi6328 committed:
Update the interaction between GUI and Logic

Hongyi6328 committed:
Change output format for GUI

Hongyi6328 committed:
Move Launcher

Hongyi6328 committed:
Update Parser.java

Hongyi6328 committed:
Merge pull request #9 from Hongyi6328/branch-A-Varargs

Update Parser.java

Hongyi6328 committed:
Rectify the exit feature

Hongyi6328 committed:
Add assertions where needed

An assertion is a statement in the Java programming language that
enables me to test my assumptions about my program.

To ensure the program at a certain stage has the expected state, we
internally check whether some conditions are met.

These assertions mainly check for null references and invalid indices.

Hongyi6328 committed:
Merge from branch-A-Assertions

At every location that is deemed important, add an insertion that
verifies some conditions are met.

These are basically critical points, such as when an instance level 
variable is assigned and will be used throughout the life cycle of 
that instance.

Hongyi6328 committed:
Improve code quality

There are too many static constants in Parser. This makes the class very
long. OOP covariants and contravariants are not considered.

Let's
* Separate those static "command suppliers" into another wrapper class.
* Add "? super" and "? extends" to generic variables if necessary.

Hongyi6328 committed:
Merge from branch-A-CodeQuality

Improve code quality

Hongyi6328 committed:
Add InstructionCommandPair class

Hongyi6328 committed:
Create CommandSupplier.java

Hongyi6328 committed:
Delete unnecessary classes

Hongyi6328 committed:
Make more use of OOP

Hongyi6328 committed:
Add Streams

Hongyi6328 committed:
Merge from branch-A-Streams

Add Streams

Hongyi6328 committed:
Add sort command

Hongyi6328 committed:
Merge from branch-C-Sort

Add sort command

Hongyi6328 committed:
Create .yml

Hongyi6328 committed:
Trigger CI

Hongyi6328 committed:
Delete a dummy line

Hongyi6328 committed:
Improve header line comments

Hongyi6328 committed:
Add Website.md

Hongyi6328 committed:
Add Ui.png

Hongyi6328 committed:
Add user guide

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update build.gradle

Hongyi6328 committed:
Add more test cases

Hongyi6328 committed:
Reformat markdown table

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Update README.md

Hongyi6328 committed:
Add better GUI

Hongyi6328 committed:
Update DialogBox.java

Hongyi6328 committed:
Set wrap text to be true
